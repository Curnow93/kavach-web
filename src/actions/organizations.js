import axios from 'axios';
import {
  ADD_ORGANIZATION,
  ADD_ORGANIZATIONS,
  SET_ORGANIZATIONS_LOADING,
  RESET_ORGANIZATIONS,
  ORGANIZATIONS_API,
  SET_SELECTED_ORGANIZATION,
} from '../constants/organizations';

export const getOrganizations = () => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .get(ORGANIZATIONS_API + '/my')
      .then((response) => {
        dispatch(addOrganizationsList(response.data));
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const getOrganization = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .get(ORGANIZATIONS_API + '/' + id)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const addOrganization = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .post(ORGANIZATIONS_API, data)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(setSelectedOrganization(response.data.id));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const updateOrganization = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .put(ORGANIZATIONS_API + '/' + data.id, data)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const deleteOrganization = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .delete(ORGANIZATIONS_API + '/' + id)
      .then(() => {
        dispatch(resetOrganizations());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const setSelectedOrganization = (id) => ({
  type: SET_SELECTED_ORGANIZATION,
  payload: id,
});

const loadingOrganizations = () => ({
  type: SET_ORGANIZATIONS_LOADING,
  payload: true,
});

const stopOrganizationsLoading = () => ({
  type: SET_ORGANIZATIONS_LOADING,
  payload: false,
});

const getOrganizationByID = (data) => ({
  type: ADD_ORGANIZATION,
  payload: data,
});

const addOrganizationsList = (data) => ({
  type: ADD_ORGANIZATIONS,
  payload: data,
});

const resetOrganizations = () => ({
  type: RESET_ORGANIZATIONS,
});
