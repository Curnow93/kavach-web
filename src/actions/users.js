import axios from 'axios';
import { ADD_USERS, SET_USERS_LOADING, RESET_USERS, USERS_API } from '../constants/users';

export const getUsers = () => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .get(USERS_API + '/' + getState().organizations.selected + '/users')
      .then((response) => {
        dispatch(addUsersList(response.data));
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const addUser = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .post(USERS_API + '/' + getState().organizations.selected + '/users', data)
      .then(() => {
        dispatch(resetUsers());
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const deleteUser = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .delete(USERS_API + '/' + getState().organizations.selected + '/users/' + id)
      .then(() => {
        dispatch(resetUsers());
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

const loadingUsers = () => ({
  type: SET_USERS_LOADING,
  payload: true,
});

const stopUsersLoading = () => ({
  type: SET_USERS_LOADING,
  payload: false,
});

const addUsersList = (data) => ({
  type: ADD_USERS,
  payload: data,
});

const resetUsers = () => ({
  type: RESET_USERS,
});
