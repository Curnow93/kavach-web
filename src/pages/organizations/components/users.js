import React from 'react';
import { Button, Select, Form, Input, Popconfirm, Table } from 'antd';
import { DeleteOutlined } from '@ant-design/icons';

function OrganizationUsers(props) {
  const [loading, setLoading] = React.useState(true);
  const [users, setUsers] = React.useState([]);

  React.useEffect(() => {
    fetch(process.env.REACT_APP_API_URL + '/organizations/' + props.organization.id + '/users')
      .then((res) => {
        if (res.status === 200) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        setUsers(res);
        setLoading(false);
      })
      .catch((err) => {
        console.log(err);
      });
  }, [props.organization.id]);

  const addUser = (values) => {
    setLoading(true);
    fetch(process.env.REACT_APP_API_URL + '/organizations/' + props.organization.id + '/users', {
      method: 'POST',
      body: JSON.stringify(values),
    })
      .then((res) => {
        if (res.status === 201) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        var newData = users;
        newData.splice(0, 0, res);
        setUsers(newData);
        setLoading(false);
      })
      .catch((err) => {
        console.log(err);
      });
  };

  const removeUser = (id) => {
    setLoading(true);
    fetch(
      process.env.REACT_APP_API_URL + '/organizations/' + props.organization.id + '/users/' + id,
      {
        method: 'DELETE',
      },
    )
      .then((res) => {
        if (res.status === 200) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        var deleteIndex = users.findIndex((item) => item.id === id);
        if (deleteIndex > -1) {
          var newData = users;
          newData.splice(deleteIndex, 1);
          setUsers(newData);
        }
        setLoading(false);
      })
      .catch((err) => {
        console.log(err);
      });
  };

  const columns = [
    {
      title: 'Name',
      key: 'name',
      render: (text, record) => record.user.first_name + ' ' + record.user.last_name,
    },
    {
      title: 'Email',
      key: 'email',
      render: (text, record) => record.user.email,
    },
    {
      title: 'Role',
      key: 'role',
      dataIndex: 'role',
    },
    {
      title: 'Action',
      key: 'action',
      render: (text, record) => (
        <Popconfirm title="Sure to delete?" onConfirm={() => removeUser(record.id)}>
          <Button icon={<DeleteOutlined />} />
        </Popconfirm>
      ),
    },
  ];

  return (
    <div>
      {props.organization.permission.role === 'owner' ? (
        <Form
          name="add_user"
          layout="inline"
          initialValues={{
            role: 'member',
          }}
          onFinish={(values) => addUser(values)}
        >
          <Form.Item
            name="email"
            placeholder="email"
            rules={[
              { required: true, message: 'Please input your title!' },
              { type: 'email', message: 'Please input valid Email!' },
            ]}
          >
            <Input />
          </Form.Item>
          <Form.Item name="role">
            <Select placeholder="role">
              <Select.Option value="owner">Owner</Select.Option>
              <Select.Option value="member">Member</Select.Option>
            </Select>
          </Form.Item>
          <Form.Item>
            <Button form="add_user" type="primary" htmlType="submit" block>
              Add User
            </Button>
          </Form.Item>
        </Form>
      ) : null}
      <Table loading={loading} pagination={false} columns={columns} dataSource={users} />
    </div>
  );
}

export default OrganizationUsers;
