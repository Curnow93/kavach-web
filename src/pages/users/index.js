import React from 'react';
import { Button, Popconfirm, Table, Form, Input, Select, Space } from 'antd';
import { useSelector, useDispatch } from 'react-redux';
import { DeleteOutlined } from '@ant-design/icons';
import { getUsers, deleteUser, addUser } from '../../actions/users';

function OrganisationUsers() {
  const [form] = Form.useForm();

  const dispatch = useDispatch();

  const { organisation, users, loading } = useSelector((state) => {
    return {
      organisation: state.organisations.details[state.organisations.selected],
      users: state.users.ids.map((id) => state.users.details[id]),
      loading: state.users.loading,
    };
  });

  const fetchUsers = React.useCallback(() => {
    dispatch(getUsers());
  }, [dispatch]);

  React.useEffect(() => {
    fetchUsers();
  }, [fetchUsers]);

  const columns = [
    {
      title: 'Name',
      key: 'name',
      render: (text, record) => record.first_name + ' ' + record.last_name,
      width: '30%',
    },
    {
      title: 'Email',
      key: 'email',
      dataIndex: 'email',
      width: '30%',
    },
    {
      title: 'Role',
      key: 'role',
      dataIndex: ['permission', 'role'],
      filters: [
        {
          text: 'Owner',
          value: 'owner',
        },
        {
          text: 'Member',
          value: 'member',
        },
      ],
      filterMultiple: false,
      onFilter: (value, record) => record.permission.role === value,
      width: '25%',
    },
    {
      title: 'Action',
      key: 'action',
      render: (text, record) => (
        <Popconfirm
          title="Sure to delete?"
          onConfirm={() => dispatch(deleteUser(record.id)).then(fetchUsers())}
        >
          <Button icon={<DeleteOutlined />} />
        </Popconfirm>
      ),
      width: '15%',
    },
  ];

  return (
    <Space direction="vertical">
      {organisation.permission.role === 'owner' ? (
        <Form
          form={form}
          name="add_user"
          layout="inline"
          initialValues={{
            role: 'member',
          }}
          onFinish={(values) =>
            dispatch(addUser(values)).then(() => {
              fetchUsers();
              form.resetFields();
            })
          }
        >
          <Form.Item
            name="email"
            placeholder="email"
            rules={[
              { required: true, message: 'Please input your title!' },
              { type: 'email', message: 'Please input valid Email!' },
            ]}
          >
            <Input />
          </Form.Item>
          <Form.Item name="role">
            <Select placeholder="role">
              <Select.Option value="owner">Owner</Select.Option>
              <Select.Option value="member">Member</Select.Option>
            </Select>
          </Form.Item>
          <Form.Item>
            <Button form="add_user" type="primary" htmlType="submit" block>
              Add User
            </Button>
          </Form.Item>
        </Form>
      ) : null}
      <Table
        rowKey={'id'}
        loading={loading}
        pagination={false}
        columns={columns}
        dataSource={users}
      />
    </Space>
  );
}

export default OrganisationUsers;
