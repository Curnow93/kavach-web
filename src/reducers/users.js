import { ADD_USERS, SET_USERS_LOADING, RESET_USERS } from '../constants/users';

const initialState = {
  ids: [],
  details: {},
  loading: true,
};

export default function usersReducer(state = initialState, action = {}) {
  switch (action.type) {
    case RESET_USERS:
      return {
        ...state,
        ids: [],
        details: {},
        loading: true,
      };
    case SET_USERS_LOADING:
      return {
        ...state,
        loading: action.payload,
      };
    case ADD_USERS:
      return {
        ...state,
        ids: action.payload.map((item) => item.id),
        details: action.payload.reduce((obj, item) => Object.assign(obj, { [item.id]: item }), {}),
      };
    default:
      return state;
  }
}
