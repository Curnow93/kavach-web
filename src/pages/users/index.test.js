import React from 'react';
import { useDispatch, useSelector, Provider } from 'react-redux';
import { act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Popconfirm } from 'antd';

import '../../matchMedia.mock';
import OrganizationUsers from './index';
import { getUsers, deleteUser, addUser } from '../../actions/users';

jest.mock('../../actions/users', () => ({
  getUsers: jest.fn(),
  addUser: jest.fn(),
  deleteUser: jest.fn(),
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Users index component', () => {
  const memberStore = mockStore({
    organizations: {
      ids: [1],
      details: {
        1: { id: 1, title: 'title', description: 'description', permission: { role: 'member' } },
      },
      loading: false,
      selected: 1,
    },
    users: {
      ids: [1],
      details: { 1: { id: 1, name: 'name', email: 'email', permission: { role: 'member' } } },
      loading: false,
    },
  });
  const ownerStore = mockStore({
    organizations: {
      ids: [1],
      details: {
        1: { id: 1, title: 'title', description: 'description', permission: { role: 'owner' } },
      },
      loading: false,
      selected: 1,
    },
    users: {
      ids: [1],
      details: {
        1: {
          id: 1,
          first_name: 'first_name',
          last_name: 'last_name',
          email: 'email',
          permission: { role: 'owner' },
        },
      },
      loading: false,
    },
  });
  beforeEach(() => {
    const mockedDispatch = jest.fn();
    mockedDispatch.mockReturnValue(Promise.resolve());
    useDispatch.mockReturnValue(mockedDispatch);
  });
  describe('snapshot testing', () => {
    it('should render the component for owner user', () => {
      let component;

      act(() => {
        component = shallow(
          <Provider store={ownerStore}>
            <OrganizationUsers />
          </Provider>,
        );
      });
      component.dive();
      expect(component).toMatchSnapshot();
    });
    it('should render the component for member user', () => {
      let component;

      act(() => {
        component = shallow(
          <Provider store={memberStore}>
            <OrganizationUsers />
          </Provider>,
        );
      });
      component.dive();
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing for owner', () => {
    let wrapper;
    beforeEach(async () => {
      await act(async () => {
        wrapper = mount(
          <Provider store={ownerStore}>
            <OrganizationUsers />
          </Provider>,
        );
      });
    });
    afterEach(() => {
      wrapper.unmount();
    });
    it('should add user', async (done) => {
      expect(getUsers).toHaveBeenCalled();
      await act(async () => {
        wrapper
          .find('FormItem')
          .at(0)
          .find('Input')
          .simulate('change', { target: { value: 'email@email.com' } });
        wrapper
          .find('FormItem')
          .at(1)
          .find('Select')
          .at(0)
          .props()
          .onChange({ target: { value: 'member' } });

        const submitButtom = wrapper.find('Button').at(0);
        submitButtom.simulate('submit');
      });
      await act(async () => {
        wrapper.update();
      });

      setTimeout(() => {
        expect(addUser).toHaveBeenCalledWith({
          email: 'email@email.com',
          role: 'member',
        });
        expect(getUsers).toHaveBeenCalled();
        done();
      });
    });
    it('should delete user', async (done) => {
      expect(getUsers).toHaveBeenCalled();
      const deleteButtom = wrapper.find('Button').at(1);
      deleteButtom.simulate('click');
      const popconfirm = wrapper.find(Popconfirm);
      popconfirm
        .findWhere((item) => item.type() === 'button' && item.text() === 'OK')
        .simulate('click');

      setTimeout(() => {
        expect(deleteUser).toHaveBeenCalledWith(1);
        expect(getUsers).toHaveBeenCalled();
        done();
      });
    });
  });
});
