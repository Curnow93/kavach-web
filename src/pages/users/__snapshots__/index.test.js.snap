// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users index component snapshot testing should render the component for member user 1`] = `ShallowWrapper {}`;

exports[`Users index component snapshot testing should render the component for owner user 1`] = `ShallowWrapper {}`;
